resources/suite-runner.mjs

```diff
@@ -89,13 +112,107 @@ export class SuiteRunner {
         this.#suiteResults.tests[test.name] = { tests: { Sync: syncTime, Async: asyncTime }, total: total };
         this.#suiteResults.total += total;
 
-        if (this.#client?.didRunTest)
-            await this.#client.didRunTest(this.#suite, test);
+        await this._updateClient(test);
     };
+
+    async _updateClient(test, suite = this.#suite) {
+        if (this.#client?.didRunTest)
+            await this.#client.didRunTest(suite, test);
+    }
 }
 
-// FIXME: implement remote steps
-class RemoteSuiteRunner extends SuiteRunner {}
+export class RemoteSuiteRunner extends SuiteRunner {
+    #appId;
+
+    get appId() {
+        return this.#appId;
+    }
+
+    set appId(id) {
+        this.#appId = id;
+    }
+
+    async run() {
+        this.postMessageCallbacks = new Map();
+        const handler = this._handlePostMessage.bind(this);
+        window.addEventListener("message", handler);
+
+        // FIXME: use this._suite in all SuiteRunner methods directly.
+        try {
+            await this._prepareSuite();
+            await this._runSuite();
+        } finally {
+            window.removeEventListener("message", handler);
+        }
+    }
+
+    async _prepareSuite() {
+        const suiteName = this.suite.name;
+        const suitePrepareStartLabel = `suite-${suiteName}-prepare-start`;
+        const suitePrepareEndLabel = `suite-${suiteName}-prepare-end`;
+
+        performance.mark(suitePrepareStartLabel);
+
+        // Wait for the app-ready message from the workload.
+        const appReadyPromise = this._subscribeOnce("app-ready");
+        await this._loadFrame(this.suite);
+        const response = await appReadyPromise;
+        await this.suite.prepare(this.page);
+        // Capture appId to pass along with messages.
+        this.appId = response?.appId;
+
+        performance.mark(suitePrepareEndLabel);
+
+        performance.measure(`suite-${suiteName}-prepare`, suitePrepareStartLabel, suitePrepareEndLabel);
+    }
+
+    async _runSuite() {
+        // Update progress bar with each completed step.
+        this._startSubscription("step-complete", async (e) => this._updateClient(e.data.test, e.data.name));
+        // Ask workload to run its own tests.
+        this.frame.contentWindow.postMessage({ id: this.appId, key: "benchmark-connector", type: "benchmark-suite", name: this.suite.config?.name || "default" }, "*");
+        // Capture metrics from the completed tests.
+        const response = await this._subscribeOnce("suite-complete");
+        this._stopSubscription("step-complete");
+
+        this.suiteResults.tests = {
+            ...this.suiteResults.tests,
+            ...response.result.tests,
+        };
+
+        this.suiteResults.total += response.result.total;
+
+        this._validateSuiteTotal();
+    }
+
+    _handlePostMessage(e) {
+        if (this.postMessageCallbacks.has(e.data.type))
```

Comment:
Can we call get here and use a local variable?
```js
if (count callback = this.postMessageCallbacks.get(e.data.type))
    callback(e);
```

Reply:
implemented above